rfc/0078-zone-aware-replica-reads.md

```diff
@@ -0,0 +1,346 @@
+# Meta
+
+| Field          | Value                    |
+|----------------|--------------------------|
+| RFC Name       | Zone-Aware Replica Reads |
+| RFC ID         | 78                       |
+| Start Date     | 2024-04-25               |
+| Owner          | Sergey Avseyev           |
+| Current Status | DRAFT                    |
+| Revision       | #1                       |
+
+# Summary
+
+This document describes changes that SDKs must implement to optimize costs of
+network traffic when the server is being deployed across multiple availability
+zones.
+
+# Motivation
+
+Existing SDKs do not take into account node availability groups of the nodes in
+the cluster. In particular, when performing `getAllReplicas` operations, the SDK
+would send `GET (0x00)` request to active vBucket and `GET_REPLICA (0x83)` to
+each of the replicas. This approach will make reading from replicas expensive
+operations if nodes of the cluster deployed in different availability zones
+(AZ).
+
+# Use Cases
+
+## No Preference
+
+This is current behaviour, where the SDK does not take into account server
+groups.
+
+![No Prefrence](figures/0078-case-1-no-preference.svg)
+
+## Selected Server Group
+
+In this case, all reads will be done only from the local availability zone. It
+is the cheapest solution, although it reduces chances of getting data from
+replicas (if the server groups are unbalanced).
+
+![Selected Server Group](figures/0078-case-2-local-only.svg)
+
+## Selected Server Group or All Available
+
+This strategy allows using all available replicas, but only in case when the
+local server group is empty. The selected server group might be empty in case
+the groups are not balanced properly on the cluster, or some nodes has been
+failed over. The previous strategy in this case would return
+`102 DocumentIrretrievable` error and refuse touching replicas from non-local
+group.
+
+Note that in this case we do not do second round trip after trying nodes from
+selected server group, but only use current topology to detect the Empty-Group
+case.
+
+![Selected Server Group or All Available](figures/0078-case-3-selected-server-group-or-all-available.svg)
```

Comment:
The inability to view these significantly impacts ability to review

Reply:
next time I will either use PNG or try to figure out something else. To me the benefit of keeping history here and removing conversion step weigh more. 

Meanwhile with couple of clicks you can see it in the fork
https://github.com/avsej/sdk-rfcs/blob/0077-zone-aware-replica-reads/rfc/0078-zone-aware-replica-reads.md

I just figured that maybe it is because of the fork, it does not show, because it insert URLs of the images, it just cannot render them as they point to this repo, not to the fork. I think it is rather github bug.

![image](https://github.com/couchbaselabs/sdk-rfcs/assets/7280/bab536f8-37ef-46bf-9165-e5caaca3b00c)